model/pod/pod.go

```diff
@@ -156,20 +184,100 @@ func (p *Pod) Remove() error {
 	)
 }
 
-// UpdateStatus updates the pod status.
+// UpdateStatus updates the pod status. If the new status is identical to the
+// current one, this is a no-op.
 func (p *Pod) UpdateStatus(s Status) error {
+	if p.Status == s {
+		return nil
+	}
+
 	byIDAndStatus := ByID(p.ID)
 	byIDAndStatus[StatusKey] = p.Status
 
+	updated := utility.BSONTime(time.Now())
+	setFields := bson.M{StatusKey: s}
+	switch s {
+	case StatusInitializing:
+		setFields[bsonutil.GetDottedKeyName(TimeInfoKey, TimeInfoInitializingKey)] = updated
+	case StatusStarting:
+		setFields[bsonutil.GetDottedKeyName(TimeInfoKey, TimeInfoStartingKey)] = updated
+	}
+
 	if err := UpdateOne(byIDAndStatus, bson.M{
-		"$set": bson.M{StatusKey: s},
+		"$set": setFields,
 	}); err != nil {
 		return err
 	}
 
 	event.LogPodStatusChanged(p.ID, string(p.Status), string(s))
 
 	p.Status = s
+	switch s {
+	case StatusInitializing:
+		p.TimeInfo.Initializing = updated
+	case StatusStarting:
+		p.TimeInfo.Starting = updated
+	}
 
 	return nil
 }
+
+// UpdateResources updates the pod resources. If the new status is identical to the current one, this is a no-op.
+func (p *Pod) UpdateResources(info ResourceInfo) error {
+	if p.Resources.Cluster == info.Cluster && p.Resources.ExternalID == info.ExternalID && p.Resources.DefinitionID == info.DefinitionID && p.checkSecretIDs(p.Resources.SecretIDs, info.SecretIDs) {
+		return nil
+	}
+
+	if info.Cluster == "" {
```

Comment:
Is there a reason that we need to keep the existing resources? I would think this should replace all the resources in case we want to do something like remove resources.

Reply:
Oh, for some reason I thought if the fields weren't set they would be nil, but the tests pass after I deleted this section, so it should be fine

Reply:
The equivalent of nil in BSON is null. But it doesn't particularly matter in this case because they're non-pointer types so they can't be nil. Also, any field with `omitempty` in the BSON struct tag will be omitted from the BSON document if they're the zero value. The zero value is nil for pointer types and the default value for value types (e.g. the empty string "" for string values).

For example, if you have this struct:
```go
ResourceInfo{
    ExternalID: "",
    DefinitionID: "task_definition_id",
    SecretIDs: nil,
    Cluster: "cluster",
}
```
The equivalent BSON document will be stored like this:
```json
{
    "definition_id": "task_definition_id",
    "cluster": "cluster"
}
```